### PR TITLE
Standardize support for Tagged and BindAddresses in Ingress Gateways

### PR DESCRIPTION
**Pull request by kyhavlov**
This updates the TaggedAddresses and BindAddresses behavior for Ingress
to match Mesh/Terminating gateways. The `consul connect envoy` command
now also allows passing an address without a port for tagged/bind
addresses.

### CODE REVIEW

File: agent/xds/listeners.go [hunk 1]
```diff
@@ -790,56 +746,49 @@ func (s *Server) makeMeshGatewayListener(name, addr string, port int, cfgSnap *p
...
+	if chain == nil || chain.IsDefault() {
+		dc := u.Datacenter
+		if dc == "" {
+			dc = cfgSnap.Datacenter
+		}
+		sni := connect.UpstreamSNI(u, "", dc, cfgSnap.Roots.TrustDomain)
+
+		useRDS = false
+		clusterName = CustomizeClusterName(sni, chain)
+	} else if cfg.Protocol == "tcp" {
```
Comment (freddygv) [inline agent/xds/listeners.go:776]:
```suggestion
		clusterName = CustomizeClusterName(sni, chain)
		
	} else if cfg.Protocol == "tcp" {
```

This block is a bit crowded, suggesting adding a newline.

File: agent/xds/listeners.go [hunk 2]
```diff
@@ -254,6 +257,12 @@ func (s *Server) listenersFromSnapshotGateway(cfgSnap *proxycfg.ConfigSnapshot,
 			if err != nil {
 				return nil, err
 			}
+		case structs.ServiceKindIngressGateway:
+			listeners, err := s.listenersFromSnapshotIngressGateway(a.Address, cfgSnap)
```
Comment (freddygv) [inline agent/xds/listeners.go:261]: Can this be renamed to match the other gateways? `s.makeIngressGatewayListeners`

File: agent/xds/listeners.go [hunk 3]
```diff
@@ -269,8 +278,9 @@ func (s *Server) listenersFromSnapshotGateway(cfgSnap *proxycfg.ConfigSnapshot,
 
 // TODO(ingress): Support configured bind addresses from similar to mesh gateways
```
Comment (freddygv) [inline agent/xds/listeners.go:279]: Outdated TODO

File: agent/xds/testdata/listeners/custom-upstream-typed-ignored-with-disco-chain.golden [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                  "stat_prefix": "upstream_prepared_query_geo-cache_tcp"
+                }
+            }
+          ]
+        }
+      ]
+    },
+    {
+      "@type": "type.googleapis.com/envoy.api.v2.Listener",
+      "name": "public_listener:0.0.0.0:9999",
```
[the remark this answers is not in the document]
Comment (crhino) [reply agent/xds/testdata/listeners/custom-upstream-typed-ignored-with-disco-chain.golden:52]: Ah no, this test is backfilling test cases for [this block](https://github.com/hashicorp/consul/blob/4277664986819bed0a886b44f13f27836d332fec/agent/xds/listeners.go#L830-L836), where we explicitly ignore the ListenerJSON if discovery chain information is available.

File: command/connect/envoy/flags.go [hunk 5]
```diff
@@ -47,8 +47,13 @@ func parseAddress(raw string) (api.ServiceAddress, error) {
...
 
 	addr, portStr, err := net.SplitHostPort(x)
+	// Error message from Go's net/ipsock.go
 	if err != nil {
-		return result, fmt.Errorf("Error parsing address %q: %v", x, err)
+		if !strings.Contains(err.Error(), "missing port in address") {
+			return result, fmt.Errorf("Error parsing address %q: %v", x, err)
+		}
+		// Use the whole input as the address if there wasn't a port.
+		addr = x
```
[the remark this answers is not in the document]
Comment (crhino) [reply command/connect/envoy/flags.go:56]: Ah yes, that makes a lot of sense, good catch!

[the remark this answers is not in the document]
Comment (crhino) [reply command/connect/envoy/flags.go:60]: done

File: command/connect/envoy/flags_test.go [hunk 6]
```diff
@@ -67,9 +67,12 @@ func TestServiceAddressValue_Set(t *testing.T) {
 			expectedValue: api.ServiceAddress{Address: "8.8.8.8", Port: 1234},
 		},
 		{
-			name:        "invalid address",
-			input:       "not-an-address",
-			expectedErr: "missing port in address",
+			name:  "address with no port",
```
[the remark this answers is not in the document]
Comment (crhino) [reply command/connect/envoy/flags_test.go:70]: done!

**Pull request merged**